lib/printer/printer.cpp

```diff
@@ -0,0 +1,261 @@
+#include "kllvm/parser/KOREParser.h"
+#include "kllvm/parser/KOREScanner.h"
+#include <kllvm/printer/printer.h>
+
+#include <map>
+
+const std::string WHITESPACE = " \n\r\t\f\v";
+
+static std::string ltrim(const std::string &s) {
+  size_t start = s.find_first_not_of(WHITESPACE);
+  return (start == std::string::npos) ? "" : s.substr(start);
+}
+
+static std::string rtrim(const std::string &s) {
+  size_t end = s.find_last_not_of(WHITESPACE);
+  return (end == std::string::npos) ? "" : s.substr(0, end + 1);
+}
+
+static std::string trim(const std::string &s) {
+  return rtrim(ltrim(s));
+}
+
+namespace kllvm {
+
+using namespace parser;
+
+std::ostream &printKORE(
+    std::ostream &os, std::string const &definitionPath,
+    std::string const &patternPath, bool hasColor, bool filterSubst) {
+  std::map<std::string, std::string> formats;
```

Comment:
If this code is going to be invoked in a running interpreter instead of a standalone executable, we need to make some modifications here. In particular, we should only build the information about the definition the first time we call this function, and then we should reuse it all in subsequent calls.

Reply:
Please also do a leak check with valgrind once you have finished refactoring the method. I know for a fact this function as it's currently written leaks the memory that is part of the definition it parses, and I want to make sure we completely address this so that multiple calls to this function won't gradually accumulate memory leaks.